src/main/scala/com/mpc/scalats/core/ScalaParser.scala

```diff
@@ -4,24 +4,139 @@ package com.mpc.scalats.core
   * Created by Milosz on 09.06.2016.
   */
 
+import org.slf4j.LoggerFactory
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
 import scala.reflect.runtime.universe._
 
+case class TypeToParse(t: Type, parent: Option[Type]) {
+
+  def typeSymbol = t.typeSymbol
+
+  def members = t.members
+
+  def typeConstructor = t.typeConstructor
+}
+
 object ScalaParser {
 
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  private[core] var alreadyExamined: mutable.Set[TypeToParse] = mutable.Set.empty
+
   import ScalaModel._
 
-  def parseCaseClasses(caseClassTypes: List[Type]): List[CaseClass] = {
-    val sealedTraitTypes = caseClassTypes.filter(isSealedTrait).distinct.map(parseSealedTrait).distinct
-    sealedTraitTypes ++
-      caseClassTypes.flatMap(getInvolvedTypes(Set.empty)).filter(isCaseClass).distinct.map(parseCaseClass(_, sealedTraitTypes)).distinct
+  def parseTypes(userInputTypes: List[Type]): List[CaseClass] = {
+
+    val initialTypes = userInputTypes.map(t => TypeToParse(t, None))
+
+    val allInvolvedTypes = initialTypes.flatMap(analyseType).distinct
+
+    allInvolvedTypes.map(parse)
+
   }
 
-  private def isSealedTrait(scalaType: Type) = {
+  private def isSealedTrait(scalaType: TypeToParse) = {
     scalaType.typeSymbol.asClass.isSealed &&
       scalaType.typeSymbol.asClass.isTrait
   }
 
-  private def parseSealedTrait(sealedTraitType: Type): CaseClass = {
+  private def analyseType(analysedType: TypeToParse): List[TypeToParse] = {
+    if (!alreadyExamined.exists(x => x.t == analysedType.t)) {
+      alreadyExamined.add(analysedType)
+      if (isSealedTrait(analysedType)) {
+        analysedType :: (directKnownSubclasses(analysedType.t).map(
+          t => TypeToParse(t, Some(analysedType.t))
+        ).flatMap(analyseType))
+
+      } else if (isCaseClass(analysedType)) {
+        analysedType :: getInvolvedTypes(analysedType).flatMap(analyseType)
+      } else {
+        Nil
+      }
+    } else {
+      Nil
+    }
+  }
+
+  private def directKnownSubclasses(tpe: Type): List[Type] = {
```

Comment:
why not take and return `TypeToParse` instead of plain `Type` ?

Reply:
No specific reason, the parent class is not needed in this method, but it would give a better look at call site, so I don't have a strong opinion here :man_shrugging: I'll change :+1: